File: src/main/java/net/atos/zac/formulieren/FormulierRuntimeService.java
package net.atos.zac.formulieren;

import static jakarta.json.JsonValue.ValueType.STRING;
import static java.time.format.DateTimeFormatter.ofPattern;
import static java.time.temporal.ChronoUnit.DAYS;
import static net.atos.zac.util.UriUtil.uuidFromURI;
import static net.atos.zac.util.time.DateTimeConverterUtil.convertToLocalDate;
import static org.apache.commons.lang3.StringUtils.*;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.Arrays;
import java.util.Comparator;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

import jakarta.inject.Inject;
import jakarta.json.*;

import org.apache.commons.lang3.BooleanUtils;
import org.apache.commons.lang3.StringUtils;
import org.flowable.task.api.Task;

import net.atos.client.zgw.drc.DrcClientService;
import net.atos.client.zgw.shared.ZGWApiService;
import net.atos.client.zgw.zrc.ZrcClientService;
import net.atos.client.zgw.zrc.model.Zaak;
import net.atos.zac.admin.ReferenceTableService;
import net.atos.zac.admin.model.ReferenceTableValue;
import net.atos.zac.app.formulieren.model.FormulierData;
import net.atos.zac.app.informatieobjecten.EnkelvoudigInformatieObjectUpdateService;
import net.atos.zac.app.task.model.RestTask;
import net.atos.zac.flowable.ZaakVariabelenService;
import net.atos.zac.flowable.task.FlowableTaskService;
import net.atos.zac.flowable.task.TaakVariabelenService;
import net.atos.zac.formulieren.model.FormulierVeldtype;
import net.atos.zac.identity.IdentityService;
import net.atos.zac.shared.helper.OpschortenZaakHelper;
import net.atos.zac.util.time.DateTimeConverterUtil;

public class FormulierRuntimeService {

    private static final DateTimeFormatter DATUM_FORMAAT = ofPattern("dd-MM-yyy");

    private static final String REDEN_ZAAK_HERVATTEN = "Zaak hervat vanuit proces";

    private static final String REFERENCE_TABLE_SEPARATOR = ";";

    private static final String DOCUMENT_SEPARATOR = ";";

    private static final String FORMIO_DEFAULT_VALUE = "defaultValue";

    private static final String FORMIO_TITLE = "title";

    private static final String AANTAL_DAGEN_VANAF_HEDEN_FORMAAT = "^[+-]\\d{1,4}$";

    @Inject
    private ZGWApiService zgwApiService;

    @Inject
    private ZrcClientService zrcClientService;

    @Inject
    private ZaakVariabelenService zaakVariabelenService;

    @Inject
    private IdentityService identityService;

    @Inject
    private ReferenceTableService referenceTableService;

    @Inject
    private OpschortenZaakHelper opschortenZaakHelper;

    @Inject
    private DrcClientService drcClientService;

    @Inject
    private EnkelvoudigInformatieObjectUpdateService enkelvoudigInformatieObjectUpdateService;

    @Inject
    private TaakVariabelenService taakVariabelenService;

    @Inject
    private FlowableTaskService flowableTaskService;

    public void renderFormulierDefinitie(final RestTask restTask) {
        final ResolveDefaultValueContext resolveDefaultValueContext = new ResolveDefaultValueContext(restTask, zrcClientService,
                zaakVariabelenService);
        restTask.getFormulierDefinitie().veldDefinities.forEach(veldDefinitie -> {
            if (isNotBlank(veldDefinitie.defaultWaarde)) {
                veldDefinitie.defaultWaarde = resolveDefaultValue(veldDefinitie.defaultWaarde, resolveDefaultValueContext);
                veldDefinitie.defaultWaarde = formatDefaultValue(veldDefinitie.defaultWaarde, veldDefinitie.veldtype);
            }
            veldDefinitie.meerkeuzeOpties = resolveMultipleChoiceOptions(veldDefinitie.meerkeuzeOpties);
        });
    }

    public JsonObject renderFormioFormulier(final RestTask restTask) {
        return copyJsonObject(restTask.getFormioFormulier(),
                new ResolveDefaultValueContext(restTask, zrcClientService, zaakVariabelenService));
    }

    public Task submit(final RestTask restTask, Task task, final Zaak zaak) {
        taakVariabelenService.setTaskinformation(task, restTask.getTaakinformatie());
        taakVariabelenService.setTaskData(task, restTask.getTaakdata());

        final var formulierData = new FormulierData(restTask.getTaakdata());

        if (formulierData.toelichting != null || formulierData.taakFataleDatum != null) {
            if (formulierData.toelichting != null) {
                task.setDescription(formulierData.toelichting);
            }
            if (formulierData.taakFataleDatum != null) {
                task.setDueDate(DateTimeConverterUtil.convertToDate(formulierData.taakFataleDatum));
            }
            task = flowableTaskService.updateTask(task);
        }
        if (formulierData.zaakOpschorten && !zaak.isOpgeschort()) {
            opschortenZaakHelper.opschortenZaak(
                    zaak,
                    DAYS.between(LocalDate.now(), convertToLocalDate(task.getDueDate())),
                    restTask.getFormulierDefinitie() != null ?
                            restTask.getFormulierDefinitie().naam :
                            restTask.getFormioFormulier().getString(FORMIO_TITLE));
        }
        if (formulierData.zaakHervatten && zaak.isOpgeschort()) {
            opschortenZaakHelper.hervattenZaak(zaak, REDEN_ZAAK_HERVATTEN);
        }
        markDocumentAsSent(formulierData);
        markDocumentAsSigned(formulierData);

        final Map<String, Object> zaakVariablen = zaakVariabelenService.readProcessZaakdata(zaak.getUuid());
        zaakVariablen.putAll(formulierData.zaakVariabelen);
        zaakVariabelenService.setZaakdata(zaak.getUuid(), zaakVariablen);

        return task;
    }

    private JsonObject copyJsonObject(
            final JsonObject jsonObject,
            final ResolveDefaultValueContext resolveDefaultValueContext
    ) {
        final var objectBuilder = Json.createObjectBuilder();
        jsonObject.entrySet().forEach(stringJsonValueEntry -> objectBuilder.add(stringJsonValueEntry.getKey(),
                copyJsonObjectValue(stringJsonValueEntry, resolveDefaultValueContext)));
        return objectBuilder.build();
    }

    private JsonValue copyJsonObjectValue(
            final Map.Entry<String, JsonValue> stringJsonValueEntry,
            final ResolveDefaultValueContext resolveDefaultValueContext
    ) {
        return stringJsonValueEntry.getValue().getValueType() == STRING &&
               stringJsonValueEntry.getKey().equals(FORMIO_DEFAULT_VALUE) ?
                       Json.createValue(resolveDefaultValue(((JsonString) stringJsonValueEntry.getValue()).getString(),
                               resolveDefaultValueContext)) :
                       copyJsonValue(stringJsonValueEntry.getValue(), resolveDefaultValueContext);
    }

    private JsonValue copyJsonValue(
            final JsonValue jsonValue,
            final ResolveDefaultValueContext resolveDefaultValueContext
    ) {
        return switch (jsonValue.getValueType()) {
            case ARRAY -> copyJsonArray(jsonValue.asJsonArray(), resolveDefaultValueContext);
            case OBJECT -> copyJsonObject(jsonValue.asJsonObject(), resolveDefaultValueContext);
            default -> jsonValue;
        };
    }

    private JsonArray copyJsonArray(
            final JsonArray jsonArray,
            final ResolveDefaultValueContext resolveDefaultValueContext
    ) {
        final var arrayBuilder = Json.createArrayBuilder();
        jsonArray.forEach(jsonValue -> arrayBuilder.add(copyJsonValue(jsonValue, resolveDefaultValueContext)));
        return arrayBuilder.build();
    }

    private String resolveDefaultValue(final String defaultValue, final ResolveDefaultValueContext context) {
        return switch (defaultValue) {
            case "TAAK:STARTDATUM" -> context.getTask().getCreatiedatumTijd().format(DATUM_FORMAAT);
            case "TAAK:FATALE_DATUM" -> context.getTask().getFataledatum().format(DATUM_FORMAAT);
            case "TAAK:GROEP" -> context.getTask().getGroep() != null ? context.getTask().getGroep().getNaam() : null;
            case "TAAK:BEHANDELAAR" -> context.getTask().getBehandelaar() != null ? context.getTask().getBehandelaar().getNaam() : null;
            case "ZAAK:STARTDATUM" -> context.getZaak().getStartdatum().format(DATUM_FORMAAT);
            case "ZAAK:FATALE_DATUM" -> context.getZaak().getUiterlijkeEinddatumAfdoening().format(DATUM_FORMAAT);
            case "ZAAK:STREEFDATUM" -> context.getZaak().getEinddatumGepland().format(DATUM_FORMAAT);
            case "ZAAK:GROEP" -> getGroepForZaakDefaultValue(context.getZaak());
            case "ZAAK:BEHANDELAAR" -> getBehandelaarForZaakDefaultValue(context.getZaak());
            default -> defaultValue.startsWith(":") ?
                    context.getZaakData().getOrDefault(defaultValue.substring(1), "").toString() :
                    defaultValue;
        };
    }

    private String getGroepForZaakDefaultValue(final Zaak zaak) {
        return zgwApiService.findGroepForZaak(zaak)
                .map(groep -> identityService.readGroup(groep.getBetrokkeneIdentificatie().getIdentificatie()).getName())
                .orElse(null);
    }

    private String getBehandelaarForZaakDefaultValue(final Zaak zaak) {
        return zgwApiService.findBehandelaarMedewerkerRoleForZaak(zaak)
                .map(behandelaar -> identityService.readUser(behandelaar.getIdentificatienummer()).getFullName())
                .orElse(null);
    }

    private String formatDefaultValue(final String defaultWaarde, final FormulierVeldtype veldtype) {
        return switch (veldtype) {
            case CHECKBOX -> formatCheckboxDefaultValue(defaultWaarde);
            case DATUM -> formatDatumDefaultValue(defaultWaarde);
            default -> defaultWaarde;
        };
    }

    private String formatCheckboxDefaultValue(final String defaultWaarde) {
        return (StringUtils.equalsIgnoreCase("ja", defaultWaarde) ||
                StringUtils.equalsIgnoreCase("true", defaultWaarde) ||
                StringUtils.equals("1", defaultWaarde)) ? BooleanUtils.TRUE : BooleanUtils.FALSE;
    }

    private String formatDatumDefaultValue(final String defaultWaarde) {
        if (defaultWaarde.matches(AANTAL_DAGEN_VANAF_HEDEN_FORMAAT)) {
            int dagen = Integer.parseInt(substring(defaultWaarde, 1));
            if (defaultWaarde.startsWith("+")) {
                return LocalDate.now().plusDays(dagen).format(DATUM_FORMAAT);
            } else {
                return LocalDate.now().minusDays(dagen).format(DATUM_FORMAAT);
            }
        } else {
            return defaultWaarde;
        }
    }

    private String resolveMultipleChoiceOptions(final String meerkeuzeOpties) {
        final var referenceTableCode = substringAfter(meerkeuzeOpties, "REF:");
        if (isNotBlank(referenceTableCode)) {
            final var referenceTable = referenceTableService.readReferenceTable(referenceTableCode);
            return referenceTable.getValues()
                    .stream()
                    .sorted(Comparator.comparingInt(ReferenceTableValue::getSortOrder))
                    .map(ReferenceTableValue::getName)
                    .collect(Collectors.joining(REFERENCE_TABLE_SEPARATOR));
        } else {
            return meerkeuzeOpties;
        }
    }

    private void markDocumentAsSent(final FormulierData formulierData) {
        if (formulierData.documentenVerzenden != null) {
            Arrays.stream(formulierData.documentenVerzenden.split(DOCUMENT_SEPARATOR))
                    .map(UUID::fromString)
                    .map(drcClientService::readEnkelvoudigInformatieobject)
                    .forEach(enkelvoudigInformatieObject -> enkelvoudigInformatieObjectUpdateService.verzendEnkelvoudigInformatieObject(
                            uuidFromURI(enkelvoudigInformatieObject.getUrl()),
                            formulierData.documentenVerzendenDatum,
                            formulierData.toelichting));
        }
    }

    private void markDocumentAsSigned(final FormulierData formulierData) {
        if (formulierData.documentenOndertekenen != null) {
            Arrays.stream(formulierData.documentenOndertekenen.split(DOCUMENT_SEPARATOR))
                    .map(UUID::fromString)
                    .map(drcClientService::readEnkelvoudigInformatieobject)
                    .filter(enkelvoudigInformatieobject -> enkelvoudigInformatieobject.getOndertekening() == null)
                    .forEach(enkelvoudigInformatieobject -> enkelvoudigInformatieObjectUpdateService.ondertekenEnkelvoudigInformatieObject(
                            uuidFromURI(enkelvoudigInformatieobject.getUrl())));
        }
    }
}

File: src/main/kotlin/net/atos/zac/formulieren/FormulierRuntimeService.kt
/*
 * SPDX-FileCopyrightText: 2023 Atos, 2024 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package net.atos.zac.formulieren

import jakarta.inject.Inject
import jakarta.json.Json
import jakarta.json.JsonArray
import jakarta.json.JsonObject
import jakarta.json.JsonString
import jakarta.json.JsonValue
import net.atos.client.zgw.drc.DrcClientService
import net.atos.client.zgw.zrc.ZrcClientService
import net.atos.client.zgw.zrc.model.Zaak
import net.atos.zac.app.formulieren.model.FormulierData
import net.atos.zac.app.formulieren.model.RESTFormulierVeldDefinitie
import net.atos.zac.flowable.ZaakVariabelenService
import net.atos.zac.flowable.task.FlowableTaskService
import net.atos.zac.flowable.task.TaakVariabelenService
import net.atos.zac.formulieren.model.FormulierVeldtype
import net.atos.zac.util.time.DateTimeConverterUtil
import nl.info.client.zgw.shared.ZGWApiService
import nl.info.client.zgw.util.extractUuid
import nl.info.zac.admin.ReferenceTableService
import nl.info.zac.admin.model.ReferenceTableValue
import nl.info.zac.app.informatieobjecten.EnkelvoudigInformatieObjectUpdateService
import nl.info.zac.app.task.model.RestTask
import nl.info.zac.identity.IdentityService
import nl.info.zac.identity.model.getFullName
import nl.info.zac.shared.helper.SuspensionZaakHelper
import org.flowable.task.api.Task
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import java.util.UUID

@Suppress("LongParameterList", "TooManyFunctions")
class FormulierRuntimeService @Inject constructor(
    private val zgwApiService: ZGWApiService,
    private val zrcClientService: ZrcClientService,
    private val zaakVariabelenService: ZaakVariabelenService,
    private val identityService: IdentityService,
    private val referenceTableService: ReferenceTableService,
    private val suspensionZaakHelper: SuspensionZaakHelper,
    private val drcClientService: DrcClientService,
    private val enkelvoudigInformatieObjectUpdateService: EnkelvoudigInformatieObjectUpdateService,
    private val taakVariabelenService: TaakVariabelenService,
    private val flowableTaskService: FlowableTaskService
) {
    companion object {
        private val DATUM_FORMAAT: DateTimeFormatter = DateTimeFormatter.ofPattern("dd-MM-yyy")

        private const val REDEN_ZAAK_HERVATTEN = "Zaak hervat vanuit proces"

        private const val REFERENCE_TABLE_SEPARATOR = ";"

        private const val DOCUMENT_SEPARATOR = ";"

        private const val FORMIO_DEFAULT_VALUE = "defaultValue"

        private const val FORMIO_TITLE = "title"

        private const val AANTAL_DAGEN_VANAF_HEDEN_FORMAAT = "^[+-]\\d{1,4}$"

        private const val DEFAULT_VALUE_JA = "ja"
        private const val DEFAULT_VALUE_TRUE = "true"
        private const val DEFAULT_VALUE_ONE = "1"
    }

    fun renderFormulierDefinitie(restTask: RestTask) =
        restTask.formulierDefinitie?.run {
            veldDefinities.forEach { fieldDefinition ->
                fieldDefinition.defaultWaarde?.let { defaultValue ->
                    setDefaultValue(defaultValue, fieldDefinition, restTask)
                }
            }
        }

    private fun setDefaultValue(
        defaultValue: String,
        fieldDefinition: RESTFormulierVeldDefinitie,
        restTask: RestTask
    ) {
        if (defaultValue.isNotBlank()) {
            fieldDefinition.defaultWaarde = resolveDefaultValue(
                defaultValue,
                ResolveDefaultValueContext(restTask, zrcClientService, zaakVariabelenService)
            )
            fieldDefinition.veldtype?.let {
                fieldDefinition.defaultWaarde = formatDefaultValue(defaultValue, it)
            }
        }
        fieldDefinition.meerkeuzeOpties?.let {
            fieldDefinition.meerkeuzeOpties = resolveMultipleChoiceOptions(it)
        }
    }

    fun renderFormioFormulier(restTask: RestTask) =
        restTask.formioFormulier?.let {
            copyJsonObject(it, ResolveDefaultValueContext(restTask, zrcClientService, zaakVariabelenService))
        }

    fun submit(restTask: RestTask, task: Task, zaak: Zaak): Task {
        var task = task
        taakVariabelenService.setTaskinformation(task, restTask.taakinformatie)
        taakVariabelenService.setTaskData(task, restTask.taakdata)

        val formulierData = FormulierData(restTask.taakdata ?: emptyMap())

        if (formulierData.toelichting != null || formulierData.taakFataleDatum != null) {
            if (formulierData.toelichting != null) {
                task.description = formulierData.toelichting
            }
            if (formulierData.taakFataleDatum != null) {
                task.dueDate = DateTimeConverterUtil.convertToDate(formulierData.taakFataleDatum)
            }
            task = flowableTaskService.updateTask(task)
        }
        if (formulierData.zaakOpschorten && !zaak.isOpgeschort) {
            suspensionZaakHelper.suspendZaak(
                zaak,
                ChronoUnit.DAYS.between(LocalDate.now(), DateTimeConverterUtil.convertToLocalDate(task.dueDate)),
                if (restTask.formulierDefinitie != null) {
                    restTask.formulierDefinitie?.naam
                } else {
                    restTask.formioFormulier?.getString(FORMIO_TITLE)
                }
            )
        }
        if (formulierData.zaakHervatten && zaak.isOpgeschort) {
            suspensionZaakHelper.resumeZaak(zaak, REDEN_ZAAK_HERVATTEN)
        }
        markDocumentAsSent(formulierData)
        markDocumentAsSigned(formulierData)

        val zaakVariablen = zaakVariabelenService.readProcessZaakdata(zaak.uuid)
        zaakVariablen.putAll(formulierData.zaakVariabelen)
        zaakVariabelenService.setZaakdata(zaak.uuid, zaakVariablen)

        return task
    }

    private fun copyJsonObject(jsonObject: JsonObject, resolveDefaultValueContext: ResolveDefaultValueContext) =
        Json.createObjectBuilder().let { objectBuilder ->
            jsonObject.entries.forEach {
                objectBuilder.add(
                    it.key,
                    copyJsonObjectValue(it, resolveDefaultValueContext)
                )
            }
            objectBuilder.build()
        }

    private fun copyJsonObjectValue(
        stringJsonValueEntry: MutableMap.MutableEntry<String, JsonValue>,
        resolveDefaultValueContext: ResolveDefaultValueContext
    ) =
        if (stringJsonValueEntry.value.valueType == JsonValue.ValueType.STRING &&
            stringJsonValueEntry.key == FORMIO_DEFAULT_VALUE
        ) {
            Json.createValue(
                resolveDefaultValue(
                    (stringJsonValueEntry.value as JsonString).string,
                    resolveDefaultValueContext
                )
            )
        } else {
            copyJsonValue(stringJsonValueEntry.value, resolveDefaultValueContext)
        }

    private fun copyJsonValue(
        jsonValue: JsonValue,
        resolveDefaultValueContext: ResolveDefaultValueContext
    ): JsonValue =
        when (jsonValue.valueType) {
            JsonValue.ValueType.ARRAY -> copyJsonArray(jsonValue.asJsonArray(), resolveDefaultValueContext)
            JsonValue.ValueType.OBJECT -> copyJsonObject(jsonValue.asJsonObject(), resolveDefaultValueContext)
            else -> jsonValue
        }

    private fun copyJsonArray(
        jsonArray: JsonArray,
        resolveDefaultValueContext: ResolveDefaultValueContext
    ) =
        Json.createArrayBuilder().let { arrayBuilder ->
            jsonArray.forEach { arrayBuilder.add(copyJsonValue(it, resolveDefaultValueContext)) }
            arrayBuilder.build()
        }

    private fun resolveDefaultValue(defaultValue: String, context: ResolveDefaultValueContext): String? =
        when (defaultValue) {
            "TAAK:STARTDATUM" -> context.task.creatiedatumTijd?.format(DATUM_FORMAAT)
            "TAAK:FATALE_DATUM" -> context.task.fataledatum?.format(DATUM_FORMAAT)
            "TAAK:GROEP" -> context.task.groep?.naam
            "TAAK:BEHANDELAAR" -> context.task.behandelaar?.naam
            "ZAAK:STARTDATUM" -> context.zaak.startdatum?.format(DATUM_FORMAAT)
            "ZAAK:FATALE_DATUM" -> context.zaak.uiterlijkeEinddatumAfdoening?.format(DATUM_FORMAAT)
            "ZAAK:STREEFDATUM" -> context.zaak.einddatumGepland?.format(DATUM_FORMAAT)
            "ZAAK:GROEP" -> getGroepForZaakDefaultValue(context.zaak)
            "ZAAK:BEHANDELAAR" -> getBehandelaarForZaakDefaultValue(context.zaak)
            else ->
                if (defaultValue.startsWith(":")) {
                    context.zaakData?.getOrDefault(defaultValue.substring(1), "").toString()
                } else {
                    defaultValue
                }
        }

    private fun getGroepForZaakDefaultValue(zaak: Zaak) =
        zgwApiService.findGroepForZaak(zaak).let { group ->
            group?.betrokkeneIdentificatie?.identificatie?.let {
                identityService.readGroup(it).name
            }
        }

    private fun getBehandelaarForZaakDefaultValue(zaak: Zaak) =
        zgwApiService.findBehandelaarMedewerkerRoleForZaak(zaak).let { behandelaar ->
            behandelaar?.getIdentificatienummer()?.let {
                identityService.readUser(it).getFullName()
            }
        }

    private fun formatDefaultValue(defaultWaarde: String, veldtype: FormulierVeldtype) =
        when (veldtype) {
            FormulierVeldtype.CHECKBOX -> formatCheckboxDefaultValue(defaultWaarde)
            FormulierVeldtype.DATUM -> formatDatumDefaultValue(defaultWaarde)
            else -> defaultWaarde
        }

    private fun formatCheckboxDefaultValue(defaultWaarde: String?) =
        if (DEFAULT_VALUE_JA.equals(defaultWaarde, ignoreCase = true) ||
            DEFAULT_VALUE_TRUE.equals(defaultWaarde, ignoreCase = true) ||
            DEFAULT_VALUE_ONE.equals(defaultWaarde, ignoreCase = true)
        ) {
            true.toString()
        } else {
            false.toString()
        }

    private fun formatDatumDefaultValue(defaultWaarde: String) =
        if (defaultWaarde.matches(AANTAL_DAGEN_VANAF_HEDEN_FORMAAT.toRegex())) {
            val dagen = defaultWaarde.substring(1).toInt()
            if (defaultWaarde.startsWith("+")) {
                LocalDate.now().plusDays(dagen.toLong()).format(DATUM_FORMAAT)
            } else {
                LocalDate.now().minusDays(dagen.toLong()).format(DATUM_FORMAAT)
            }
        } else {
            defaultWaarde
        }

    private fun resolveMultipleChoiceOptions(multipleChoiceOptions: String) =
        multipleChoiceOptions.substringAfter("REF:").let { referenceTableCode ->
            if (referenceTableCode.isNotBlank()) {
                referenceTableService.readReferenceTable(referenceTableCode).let { referenceTableValue ->
                    referenceTableValue.values
                        .sortedWith(Comparator.comparingInt(ReferenceTableValue::sortOrder))
                        .joinToString(REFERENCE_TABLE_SEPARATOR) { it.name }
                }
            } else {
                multipleChoiceOptions
            }
        }

    private fun markDocumentAsSent(formulierData: FormulierData) =
        formulierData.documentenVerzenden?.let { documentsToMark ->
            documentsToMark.split(DOCUMENT_SEPARATOR.toRegex())
                .dropLastWhile { it.isEmpty() }
                .map { UUID.fromString(it) }
                .map { drcClientService.readEnkelvoudigInformatieobject(it) }
                .forEach {
                    enkelvoudigInformatieObjectUpdateService.verzendEnkelvoudigInformatieObject(
                        it.url.extractUuid(),
                        formulierData.documentenVerzendenDatum,
                        formulierData.toelichting
                    )
                }
        }

    private fun markDocumentAsSigned(formulierData: FormulierData) =
        formulierData.documentenOndertekenen?.let { documentenOndertekenen ->
            documentenOndertekenen.split(DOCUMENT_SEPARATOR.toRegex()).dropLastWhile { it.isEmpty() }
                .map { UUID.fromString(it) }
                .map { drcClientService.readEnkelvoudigInformatieobject(it) }
                .filter { it.ondertekening == null }
                .forEach {
                    enkelvoudigInformatieObjectUpdateService.ondertekenEnkelvoudigInformatieObject(
                        it.url.extractUuid()
                    )
                }
        }
}
